File: Tree/113_PathSumII.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-19 10:25:40
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> pathList;
        vector<int> path;
        findPath(root, sum, path, pathList);
        return pathList;
    }

    void findPath(TreeNode* root, int sum, vector<int> path, vector<vector<int>> &pathList){
        if(root==NULL)  return;
        path.push_back(root->val);
        if(root->left == NULL && root->right==NULL && root->val==sum){
            pathList.push_back(path);
        }
        findPath(root->left, sum-root->val, path, pathList);
        findPath(root->right, sum-root->val, path, pathList);
        path.pop_back();
    }
};

/*
[]
0
[1,2,3,4,null,6,7,5,8]
15
[1,2,2,3,3,3,3]
6
*/

File: Tree/113_PathSumII.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        if not root:
            return []
        paths_list = []
        if not root.left and not root.right:
            if root.val == sum:
                paths_list.append([root.val])
            return paths_list

        if root.left:
            l_paths = self.pathSum(root.left, sum-root.val)
            # There are paths along root.left
            if l_paths:
                for path in l_paths:
                    one_path = [root.val]
                    one_path.extend(path)
                    paths_list.append(one_path)

        if root.right:
            r_paths = self.pathSum(root.right, sum-root.val)
            # There are paths along root.right
            if r_paths:
                for path in r_paths:
                    one_path = [root.val]
                    one_path.extend(path)
                    paths_list.append(one_path)
        return paths_list


# Pythonic way.  So short and beautiful!
class Solution_2(object):
    def pathSum(self, root, sum):
        if not root:
            return []
        if not root.left and not root.right and sum == root.val:
            return [[root.val]]
        tmp = (self.pathSum(root.left, sum-root.val) +
               self.pathSum(root.right, sum-root.val))
        return [[root.val] + i for i in tmp]


"""
[]
0
[1,2,3,4,null,6,7,5,8]
15
[1,2,2,3,3,3,3]
6
"""
